### PR TITLE
[wasm][debugger] Don't show the wasm functions on callstack when debugging managed code

### LINKED ISSUE #63466: [wasm][debugger] Show a better stack trace for final user
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1425707?src=WorkItemMention&src-action=artifact_link
As discussed with Larry, we would like to:
- [ ] We would like to choose if we will show the wasm functions or not in the callstack (maybe based on runtime version, if runtime is debug we should show, otherwise we shouldn't?)
- [ ] If the runtime is on debug and we have debug symbols, work well with the google dwarf wasm extension.

### PR DESCRIPTION
**Pull request by thaystg**
This PR is what I would like to backport to 6.0.

On .net7 my plan is to check if there is dwarf symbols or create something to enable it, the callstack with wasm functions are useful when we want to debug the runtime code.

But I would like to merge this version and backport to make callstack have the same behavior that we had on .net5.

In the first time that we run the Blazor project we get this:
`"http://localhost:5174/_framework/dotnet.wasm"` in the URL field of the original callstack that came from the browser.
After stop and running again we get this:
`"wasm://wasm/00945abe"`

That is why the user was complaining about seeing different callstack on first and second execution.

Fixes #63466

### CODE REVIEW

**Review by ilonatommy — APPROVED** [review]

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs [hunk 1]
```diff
@@ -961,7 +961,9 @@ private async Task<bool> SendCallStack(SessionId sessionId, ExecutionContext con
                 string function_name = frame["functionName"]?.Value<string>();
                 string url = frame["url"]?.Value<string>();
                 if (!(function_name.StartsWith("wasm-function", StringComparison.Ordinal) ||
-                        url.StartsWith("wasm://wasm/", StringComparison.Ordinal) || function_name == "_mono_wasm_fire_debugger_agent_message"))
+                        url.StartsWith("wasm://wasm/", StringComparison.Ordinal) ||
```
Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:964]: the script parsed event has another different condition, should they be the same?

File: src/mono/wasm/debugger/DebuggerTestSuite/BreakpointTests.cs [hunk 2]
```diff
@@ -155,6 +155,9 @@ public async Task CreateGoodBreakpointAndHit()
                     Assert.Equal("object", scope["object"]["type"]);
                     CheckLocation("dotnet://debugger-test.dll/debugger-test.cs", 8, 4, scripts, scope["startLocation"]);
                     CheckLocation("dotnet://debugger-test.dll/debugger-test.cs", 14, 4, scripts, scope["endLocation"]);
+
+                    var not_wasm_frame = pause_location["callFrames"][2];
+                    Assert.Contains("dotnet.js", not_wasm_frame["url"].Value<string>());
```
Comment (radical) [inline src/mono/wasm/debugger/DebuggerTestSuite/BreakpointTests.cs:160]: Should the test scan all the frames to check for `wasm://` also?

Comment (radical) [reply src/mono/wasm/debugger/DebuggerTestSuite/BreakpointTests.cs:160]: or whatever the condition that is used in the proxy

**Comment by radical** [comment]
Will this handle the cases - with, and without debug symbols? Also, it would be useful to have a way to turn these back on.

**Comment by thaystg** [comment]
> Will this handle the cases - with, and without debug symbols? Also, it would be useful to have a way to turn these back on.

In this PR we are ignoring wasm it doesn't matter if it is with or without symbols. And this is what I want to be the behavior on .net6.

For .net7 I will consider if it's with symbols and also support debugging something like a "mixed" mode, where we can debug the managed code, and using the DWARF extension debug the native code. This was already tested by me and it works :)

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs (hunk 1 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:965]: done.

**Comment by radical** [comment]
> In this PR we are ignoring wasm it doesn't matter if it is with or without symbols. And this is what I want to be the behavior on .net6.

I meant, does the url follow the same rules (starts with `wasm://`, or ending with `.wasm`) in case of building with, or without symbols, but I guess that would be the case of AOT, not here. Ignore my comment :)

Having a way to not filter these frames out would be useful for those working on the runtime.

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs (hunk 1 above)
Comment (radical) [reply src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:965]: Maybe convert to a method, so the two places stay in sync? `IsAWasmUrl`?

**Comment by thaystg** [comment]
> > In this PR we are ignoring wasm it doesn't matter if it is with or without symbols. And this is what I want to be the behavior on .net6.
> 
> I meant, does the url follow the same rules (starts with `wasm://`, or ending with `.wasm`) in case of building with, or without symbols, but I guess that would be the case of AOT, not here. Ignore my comment :)
> 
> Having a way to not filter these frames out would be useful for those working on the runtime.

Yes, I'll implement a way to not filter these frames for net7, do you think I should find a solution for it for .net6 also?

**Comment by thaystg** [comment]
@lambdageek , As we have already discussed offline we would like to have a way to enable and disable showing the wasm functions on callstack, do you think this is important for .net6?
Or in .net6 I can just hide the wasm functions and for .net7 I implement something that we can enable and disable it and also showing wasm functions based on having dwarf information?

**Comment by lambdageek** [comment]
> @lambdageek , As we have already discussed offline we would like to have a way to enable and disable showing the wasm functions on callstack, do you think this is important for .net6? 

We talked offline.  We can't think of a scenario where seeing the mangled wasm function names (which are not the C function symbols) is useful either for end users or for runtime developers.  (even if linking native libraries, or using AOT) So it seems ok to disable completely.

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs [hunk 3]
```diff
@@ -237,6 +237,7 @@ protected override async Task<bool> AcceptEvent(SessionId sessionId, string meth
                         {
                             case var _ when url == "":
                             case var _ when url.StartsWith("wasm://", StringComparison.Ordinal):
```
Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:239]: is there a reason we're using`url.StartsWith("wasm://"` here and `url.StartsWith("wasm://wasm/",` in SendCallStack?

Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:239]: Done.

**Review by lewing — APPROVED** [review]

**Pull request merged**